### PR TITLE
Inbound refactor

### PR DESCRIPTION
**Pull request by sbenedicadb**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by swarna04** [comment]
Steve, thanks for all the work here. I've provided my feedback, some of which bears questions/items we might need to address in near future.

**Review by swarna04 — CHANGES_REQUESTED** [review]

File: AEPMessaging/Tests/Resources/cachedProposition.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+                },
+                "decisionProvider": "AJO",
+                "activity": {
+                    "id": "4529e0b2-e1c4-4f14-8fc3-7e46263f7524#4e1235ba-b202-457e-a54d-aa020cd9e8b0"
+                }
+            },
+            "id": "ac22a428-4794-46f0-a731-728cbfa70a0e",
+            "items": [
+                {
+                    "data": {
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Tests/Resources/cachedProposition.json:17]: this is a v1 iam handle and it's in the expected format 👍

File: AEPMessaging/Tests/Resources/feedProposition.json [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                                ]
+                            }
+                        },
+                        "consequences": [
+                            {
+                                "id": "183639c4-cb37-458e-a8ef-4e130d767ebf",
+                                "type": "schema",
+                                "detail": {
+                                    "id": "183639c4-cb37-458e-a8ef-4e130d767ebf",
+                                    "schema": "https://ns.adobe.com/personalization/inbound/feed-item",
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Tests/Resources/feedProposition.json:56]: one of the many tests to be updated 😄

File: AEPMessaging/Tests/Resources/inboundConsequenceDetail.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+    "id": "183639c4-cb37-458e-a8ef-4e130d767ebf",
+    "schema": "https://ns.adobe.com/personalization/inbound/feed-item",
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Tests/Resources/inboundConsequenceDetail.json:3]: nope.  i'll remove it.

File: AEPMessaging/Sources/MessagingPropositionItem.swift [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    /// contains value for `id` in JSON
+    public let propositionId: String
+
+    /// PropositionItem schema string
+    /// contains value for `schema` in JSON
+    public let schema: SchemaType
+
+    /// PropositionItem data as dictionary
+    /// contains value for `data` in JSON
+    public let propositionData: [String: Any]?
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Sources/MessagingPropositionItem.swift:37]: will rename to `itemId` and `itemData` for brevity

File: AEPMessaging/Sources/schemas/FeedItemSchemaData.swift [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+            try container.encode(content as? String, forKey: .content)
+        }
+        try container.encode(publishedDate, forKey: .publishedDate)
+        try container.encode(expiryDate, forKey: .expiryDate)
+        try container.encode(AnyCodable.from(dictionary: meta), forKey: .meta)
+    }
+}
+
+extension FeedItemSchemaData {
+    static func getEmpty() -> FeedItemSchemaData {
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Sources/schemas/FeedItemSchemaData.swift:75]: testing only - added a comment as such

File: AEPMessaging/Sources/ParsedPropositions.swift [hunk 6]
```diff
@@ -35,55 +35,84 @@ struct ParsedPropositions {
...
+                        continue
                     }
+                    guard let consequence = parsedRules.first?.consequences.first,
+                          let schemaConsequence = MessagingPropositionItem.fromRuleConsequence(consequence) else {
+                        continue
+                    }
+                    
+                    // handle these schemas when they're embedded in ruleset-item schemas:
+                    // a. in-app schema consequences get persisted to disk, cached for reporting, and added to rules that need to be updated
+                    //    i. default-content schema consequences are treated like in-app at a proposition level
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Sources/ParsedPropositions.swift:69]: yes - 100%.  

results from ongoing rules engine discussions will determine path forward for supporting that.

File: AEPMessaging/Sources/Messaging.swift [hunk 7]
```diff
@@ -265,17 +267,60 @@ public class Messaging: NSObject, Extension {
...
+                for propositionItem in propositionItemsArray {
+                    // TODO: REVERT THIS
+                    guard let propositionInfo = propositionInfo.first?.value else { //propositionInfo[propositionItem.propositionId] else {
+                        continue
+                    }
+                    
+                    // get proposition that this item belongs to
+                    let proposition = MessagingProposition(
+                        // TODO: REVERT THIS TOO
+                        uniqueId: UUID().uuidString, // propositionInfo.id,
```
[the remark this answers is not in the document]
Comment (sbenedicadb) [reply AEPMessaging/Sources/Messaging.swift:327]: waiting for XAS updates to be deployed.  leaving these in for now as they enable e2e testing.

File: AEPMessaging/Sources/schemas/FeedItemSchemaData.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (swarna04) [reply AEPMessaging/Sources/schemas/FeedItemSchemaData.swift:76]: nit: We should use conditional compilation for these #if DEBUG...#endif, just the way we are doing for some other methods needed for testing purposes only.

File: AEPMessaging/Sources/schemas/InAppSchemaData.swift [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        try container.encode(publishedDate, forKey: .publishedDate)
+        try container.encode(expiryDate, forKey: .expiryDate)
+        try container.encode(AnyCodable.from(dictionary: meta), forKey: .meta)
+        try container.encode(AnyCodable.from(dictionary: mobileParameters), forKey: .mobileParameters)
+        try container.encode(AnyCodable.from(dictionary: webParameters), forKey: .webParameters)
+        try container.encode(remoteAssets, forKey: .remoteAssets)
+    }
+}
+
+extension InAppSchemaData {
```
Comment (swarna04) [inline AEPMessaging/Sources/schemas/InAppSchemaData.swift:79]: 👍

File: AEPMessaging/Sources/MessagingPropositionItem.swift [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        try container.encode(AnyCodable.from(dictionary: propositionData), forKey: .data)
+    }
+}
+
+public extension MessagingPropositionItem {
+    static func fromRuleConsequence(_ consequence: RuleConsequence) -> MessagingPropositionItem? {
+        guard let detailsData = try? JSONSerialization.data(withJSONObject: consequence.details, options: .prettyPrinted) else {
+            return nil
+        }
+        return try? JSONDecoder().decode(MessagingPropositionItem.self, from: detailsData)
```
[the remark this answers is not in the document]
Comment (swarna04) [reply AEPMessaging/Sources/MessagingPropositionItem.swift:77]: This still needs to be addressed, we do not want to lose the original item Id (by replacing with schema consequence Id) as we shall need to echo it back during item level tracking.

**Review by swarna04 — APPROVED** [review]

**Pull request merged**